File: app/src/main/java/com/storehop/app/data/dao/ItemDao.kt
package com.storehop.app.data.dao

import androidx.room.Dao
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import androidx.room.Upsert
import com.storehop.app.data.db.relations.ItemWithCategoryAndStores
import com.storehop.app.data.entity.Item
import kotlinx.coroutines.flow.Flow

/**
 * v0.7.0 access scope: queries filter by `householdId` (not `userId`).
 * See [StoreDao] for the rationale.
 */
@Dao
interface ItemDao {

    @Transaction
    @Query(
        """
        SELECT * FROM items
        WHERE householdId = :householdId AND deletedAt IS NULL
        ORDER BY name COLLATE NOCASE
        """,
    )
    fun observeAll(householdId: String): Flow<List<ItemWithCategoryAndStores>>

    @Query(
        """
        SELECT * FROM items
        WHERE householdId = :householdId AND deletedAt IS NULL AND isNeeded = 1
        ORDER BY name COLLATE NOCASE
        """,
    )
    fun observeNeeded(householdId: String): Flow<List<Item>>

    @Transaction
    @Query(
        """
        SELECT * FROM items
        WHERE id = :id AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    fun observeById(householdId: String, id: String): Flow<ItemWithCategoryAndStores?>

    @Upsert
    suspend fun upsert(item: Item)

    /**
     * Batch upsert for the pull side. Mappers stamp `pendingSync = false` so
     * pulled rows don't immediately re-push. Called inside [PullWriteDao]'s
     * single transaction.
     */
    @Upsert
    suspend fun upsertFromCloud(rows: List<Item>)

    @Query(
        """
        UPDATE items
        SET deletedAt = :now, updatedAt = :now, pendingSync = 1
        WHERE id = :id AND householdId = :householdId
        """,
    )
    suspend fun softDelete(householdId: String, id: String, now: Long)

    /** Tombstone-aware lookup for the undo path. */
    @Query("SELECT * FROM items WHERE id = :id AND householdId = :householdId LIMIT 1")
    suspend fun findAnyById(householdId: String, id: String): Item?

    /**
     * Live, case-insensitive name lookup. Used by the CSV import duplicate
     * guard: if an alive item with the same name already exists, the import
     * skips that row to preserve the user's existing data per the
     * "don't erase anything" hard constraint. Items aren't UNIQUE on name
     * in the schema (you can have a "Milk" from Lidl and another "Milk"
     * from Mimosa), so this is a soft dedup just for import.
     */
    @Query(
        """
        SELECT * FROM items
        WHERE householdId = :householdId
          AND name = :name COLLATE NOCASE
          AND deletedAt IS NULL
        LIMIT 1
        """,
    )
    suspend fun findByName(householdId: String, name: String): Item?

    /** Inverse of [softDelete]: clears `deletedAt`, re-flags pendingSync. */
    @Query(
        """
        UPDATE items
        SET deletedAt = NULL, updatedAt = :now, pendingSync = 1
        WHERE id = :id AND householdId = :householdId
        """,
    )
    suspend fun restoreFromTombstone(householdId: String, id: String, now: Long)

    @Query(
        """
        UPDATE items
        SET isNeeded = 0, lastPurchasedAt = :now, updatedAt = :now, pendingSync = 1
        WHERE id = :id AND householdId = :householdId
        """,
    )
    suspend fun markPurchased(householdId: String, id: String, now: Long)

    /**
     * Restore an item to the "needed" list. Used by the Shop-at-Store screen
     * when the user taps a struck-through purchased staple to put it back on
     * the list (un-checks it). Does not touch `lastPurchasedAt` -- the prior
     * purchase still happened.
     */
    @Query(
        """
        UPDATE items
        SET isNeeded = 1, updatedAt = :now, pendingSync = 1
        WHERE id = :id AND householdId = :householdId
        """,
    )
    suspend fun markNeeded(householdId: String, id: String, now: Long)

    /**
     * Clear `categoryId` on every live item that points at a category that's
     * being soft-deleted. Used by the category cascade — without this, items
     * would still resolve their @Relation join to the (tombstoned) category
     * row. Setting categoryId = NULL makes the LEFT JOIN return null cleanly.
     */
    @Query(
        """
        UPDATE items
        SET categoryId = NULL, updatedAt = :now, pendingSync = 1
        WHERE categoryId = :categoryId AND householdId = :householdId AND deletedAt IS NULL
        """,
    )
    suspend fun clearCategoryReferences(householdId: String, categoryId: String, now: Long)

    /**
     * Inverse of [clearCategoryReferences]: re-link items to their previous
     * category when the user undoes a category soft-delete. Identifies
     * affected items by `updatedAt = :clearedAt AND categoryId IS NULL`,
     * which is the unique stamp the cascade-clear left behind. Not a hard
     * guarantee (an unrelated item update at the exact same ms could match)
     * but the undo window is seconds, so the collision risk is tiny.
     */
    @Query(
        """
        UPDATE items
        SET categoryId = :categoryId, updatedAt = :now, pendingSync = 1
        WHERE householdId = :householdId AND categoryId IS NULL
          AND updatedAt = :clearedAt AND deletedAt IS NULL
        """,
    )
    suspend fun restoreCategoryReferences(householdId: String, categoryId: String, clearedAt: Long, now: Long)

    @Query("SELECT * FROM items WHERE householdId = :householdId AND pendingSync = 1")
    fun observePendingPush(householdId: String): Flow<List<Item>>

    /** v0.7.1: row-count of pending pushes for the Force-sync-now UX. */
    @Query("SELECT COUNT(*) FROM items WHERE householdId = :householdId AND pendingSync = 1")
    fun countPendingPush(householdId: String): Flow<Int>

    /**
     * v0.8.0.4: snapshot of primary keys for rows with pendingSync = 1
     * in this household. Read by [PullWriteDao.replaceAllForUid] to
     * filter cloud rows that would otherwise resurrect local edits
     * waiting to be pushed. See Mike's write-revert bug for the
     * canonical case.
     */
    @Query("SELECT id FROM items WHERE householdId = :householdId AND pendingSync = 1")
    suspend fun pendingPushIds(householdId: String): List<String>

    @Query("UPDATE items SET pendingSync = 0 WHERE id = :id AND householdId = :householdId")
    suspend fun markPushed(householdId: String, id: String)
}

File: ios/Storehop/Data/DAO/ItemDao.swift
import Foundation
import GRDB

/// v0.7.0 access scope: queries filter by `householdId` (not `userId`).
/// See `StoreDao` for the rationale.
struct ItemDao: Sendable {
    let writer: any DatabaseWriter

    // MARK: - Reactive

    func observeAll(householdId: String) -> AsyncValueObservation<[ItemWithCategoryAndStores]> {
        ValueObservation
            .tracking { db in
                try ItemWithCategoryAndStores.fetchAll(db, householdId: householdId)
            }
            .values(in: writer)
    }

    func observeNeeded(householdId: String) -> AsyncValueObservation<[Item]> {
        ValueObservation
            .tracking { db in
                try Item.fetchAll(db, sql: """
                    SELECT * FROM items
                    WHERE householdId = ? AND deletedAt IS NULL AND isNeeded = 1
                    ORDER BY name COLLATE NOCASE
                    """, arguments: [householdId])
            }
            .values(in: writer)
    }

    func observeById(householdId: String, id: String) -> AsyncValueObservation<ItemWithCategoryAndStores?> {
        ValueObservation
            .tracking { db in
                try ItemWithCategoryAndStores.fetch(db, householdId: householdId, id: id)
            }
            .values(in: writer)
    }

    func observePendingPush(householdId: String) -> AsyncValueObservation<[Item]> {
        ValueObservation
            .tracking { db in
                try Item.fetchAll(db, sql: "SELECT * FROM items WHERE householdId = ? AND pendingSync = 1", arguments: [householdId])
            }
            .values(in: writer)
    }

    /// v0.7.1: row-count of pending pushes for the Force-sync-now UX.
    func countPendingPush(householdId: String) -> AsyncValueObservation<Int> {
        ValueObservation
            .tracking { db in
                try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM items WHERE householdId = ? AND pendingSync = 1", arguments: [householdId]) ?? 0
            }
            .values(in: writer)
    }

    // MARK: - Snapshot reads

    func findAnyById(householdId: String, id: String) async throws -> Item? {
        try await writer.read { db in try Self.findAnyById(on: db, householdId: householdId, id: id) }
    }

    static func findAnyById(on db: Database, householdId: String, id: String) throws -> Item? {
        try Item.fetchOne(db, sql: "SELECT * FROM items WHERE id = ? AND householdId = ? LIMIT 1", arguments: [id, householdId])
    }

    /// Snapshot of "live item with relations" for use inside a repo
    /// transaction (e.g. updateItem needs the current row to preserve
    /// isNeeded/lastPurchasedAt/createdAt).
    static func findLiveById(on db: Database, householdId: String, id: String) throws -> Item? {
        try Item.fetchOne(db, sql: """
            SELECT * FROM items
            WHERE id = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [id, householdId])
    }

    /// Live, case-insensitive name lookup. Used by the Shop-at-Store
    /// QuickAdd flow to dedupe before creating: typing "Milk" when "Milk"
    /// already exists must re-tag the existing item, not duplicate it.
    /// Mirrors Android's `ItemDao.findByName` (`COLLATE NOCASE`).
    static func findByName(on db: Database, householdId: String, name: String) throws -> Item? {
        try Item.fetchOne(db, sql: """
            SELECT * FROM items
            WHERE householdId = ?
              AND name = ? COLLATE NOCASE
              AND deletedAt IS NULL
            LIMIT 1
            """, arguments: [householdId, name])
    }

    // MARK: - Writes

    func upsert(_ item: Item) async throws {
        try await writer.write { db in
            var copy = item
            try copy.upsert(db)
        }
    }

    func upsertFromCloud(_ rows: [Item], on db: Database) throws {
        for row in rows {
            var copy = row
            try copy.upsert(db)
        }
    }

    func softDelete(householdId: String, id: String, now: Int64) async throws {
        try await writer.write { db in try Self.softDelete(on: db, householdId: householdId, id: id, now: now) }
    }

    static func softDelete(on db: Database, householdId: String, id: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE items
            SET deletedAt = ?, updatedAt = ?, pendingSync = 1
            WHERE id = ? AND householdId = ?
            """, arguments: [now, now, id, householdId])
    }

    func restoreFromTombstone(householdId: String, id: String, now: Int64) async throws {
        try await writer.write { db in try Self.restoreFromTombstone(on: db, householdId: householdId, id: id, now: now) }
    }

    static func restoreFromTombstone(on db: Database, householdId: String, id: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE items
            SET deletedAt = NULL, updatedAt = ?, pendingSync = 1
            WHERE id = ? AND householdId = ?
            """, arguments: [now, id, householdId])
    }

    /// Item-level mark-purchased. Vestigial after v4→v5; use
    /// `ItemStoreXrefDao.markPurchasedAcrossAllStores` for the user-facing
    /// flow so per-store state cascades correctly.
    func markPurchased(householdId: String, id: String, now: Int64) async throws {
        try await writer.write { db in
            try db.execute(sql: """
                UPDATE items
                SET isNeeded = 0, lastPurchasedAt = ?, updatedAt = ?, pendingSync = 1
                WHERE id = ? AND householdId = ?
                """, arguments: [now, now, id, householdId])
        }
    }

    func markNeeded(householdId: String, id: String, now: Int64) async throws {
        try await writer.write { db in
            try db.execute(sql: """
                UPDATE items
                SET isNeeded = 1, updatedAt = ?, pendingSync = 1
                WHERE id = ? AND householdId = ?
                """, arguments: [now, id, householdId])
        }
    }

    func clearCategoryReferences(householdId: String, categoryId: String, now: Int64) async throws {
        try await writer.write { db in
            try Self.clearCategoryReferences(on: db, householdId: householdId, categoryId: categoryId, now: now)
        }
    }

    static func clearCategoryReferences(on db: Database, householdId: String, categoryId: String, now: Int64) throws {
        try db.execute(sql: """
            UPDATE items
            SET categoryId = NULL, updatedAt = ?, pendingSync = 1
            WHERE categoryId = ? AND householdId = ? AND deletedAt IS NULL
            """, arguments: [now, categoryId, householdId])
    }

    /// v0.9 "Buy Today!": set (or clear) the transient urgency flag on an item.
    /// Cleared automatically on purchase; set from the item form. Idempotent.
    static func setBuyToday(on db: Database, householdId: String, id: String, value: Bool, now: Int64) throws {
        try db.execute(sql: """
            UPDATE items
            SET isBuyToday = ?, updatedAt = ?, pendingSync = 1
            WHERE id = ? AND householdId = ?
            """, arguments: [value, now, id, householdId])
    }

    func restoreCategoryReferences(householdId: String, categoryId: String, clearedAt: Int64, now: Int64) async throws {
        try await writer.write { db in
            try Self.restoreCategoryReferences(on: db, householdId: householdId, categoryId: categoryId, clearedAt: clearedAt, now: now)
        }
    }

    static func restoreCategoryReferences(on db: Database, householdId: String, categoryId: String, clearedAt: Int64, now: Int64) throws {
        try db.execute(sql: """
            UPDATE items
            SET categoryId = ?, updatedAt = ?, pendingSync = 1
            WHERE householdId = ? AND categoryId IS NULL
              AND updatedAt = ? AND deletedAt IS NULL
            """, arguments: [categoryId, now, householdId, clearedAt])
    }

    func markPushed(householdId: String, id: String) async throws {
        try await writer.write { db in
            try db.execute(sql: "UPDATE items SET pendingSync = 0 WHERE id = ? AND householdId = ?", arguments: [id, householdId])
        }
    }
}
